### PR TITLE
Remove some code assumptions of JSON-RPC response format

### PR DESCRIPTION
**Pull request by eddyashton**
Creating this as a draft PR while I work out which bits of the test matrix I've missed, and do some tidying.

Core goal is that handlers shouldn't be constructing and returning JSON response objects, they should be filling in the response fields on the context and this later constructs the full response (in JSON-RPC or HTTP, or eventually flatbuffer-over-websocket). So `return jsonrpc::error(code, msg)` becomes `rpc_ctx.set_response_error(code, msg)`. There are convenience wrappers around handler definitions so that they can have a clean definition - see `handleradapter.h`.

While doing this I've started to break the connection between `Frontend`s and the set of handlers they support, to reduce the total code in `frontend.h`. Eventually we should be able to support multiple `HandlerRegistry`s with their own dispatch methods, independently of the shared execution logic in `Frontend`.

This is still sending JSON-RPC as the payload in HTTP mode. Next time round I will add a `HttpRpcContext` as an alternative with; a generic payload, dispatch-related fields requested from the object, and removing hard-coded JSON-RPC error codes

### CODE REVIEW

File: src/apps/jsgeneric/jsgeneric.cpp [hunk 1]
```diff
@@ -64,24 +73,45 @@ namespace ccfapp
...
+            auto err_message_it = err_it->find("message");
+            if (err_message_it != err_it->end())
+            {
+              msg = *err_message_it;
+            }
+          }
+
+          args.rpc_ctx->set_response_error(err_code, msg);
+          return;
         }
```
Comment (eddyashton) [inline src/apps/jsgeneric/jsgeneric.cpp]: To clarify the whole section above: The Lua and JS handlers currently need to give a return value and indicate whether this was a success or an error message. They should return an object, either `{result = ...}` or `{error = ...}`, we use the name of the field to work out whether it was an error, and use the `...` as the corresponding payload. In future, we may want to expose the context itself to the interpreter, so the handler can parse the request then set response headers and payload manually.

**Comment by cimetrics** [comment]
abstract_context@[3868 aka 20200120.3](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3868&view=results) vs master ewma over 20 builds from [3742](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3742&view=results) to [3864](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3864&view=results)
images

**Comment by cimetrics** [comment]
abstract_context@[3885 aka 20200120.12](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3885&view=results) vs master ewma over 20 builds from [3743](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3743&view=results) to [3880](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3880&view=results)
images

File: src/enclave/enclavetypes.h [hunk 2]
```diff
@@ -2,136 +2,12 @@
...
-      throw std::logic_error(fmt::format("Failed to unpack: {}", rpc.dump()));
-    }
-
-    parse_rpc_context(rpc_ctx, rpc);
-    rpc_ctx.raw = packed;
-
-    return rpc_ctx;
-  }
-
   class AbstractRPCResponder
```
Comment (jumaffre) [inline src/enclave/enclavetypes.h]: Perhaps rename `enclavetypes.h` to something more forwarder-specific?

File: src/enclave/rpccontext.h [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+    int code;
+    std::string msg;
+  };
+
+  struct RpcResponse
+  {
+    std::variant<ErrorDetails, nlohmann::json> result;
+  };
+
+  class RpcContext
```
Comment (jumaffre) [inline src/enclave/rpccontext.h]: Minor: it seems that we've got some inconsistencies between `Rpc` and `RPC` - I'm happy with either but we should agree on one.

File: src/node/rpc/handlerregistry.h [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+      Forwardable forwardable;
+      bool execute_locally = false;
+    };
+
+  protected:
+    std::optional<Handler> default_handler;
+    std::unordered_map<std::string, Handler> handlers;
+
+    kv::Consensus* consensus = nullptr;
+    kv::TxHistory* history = nullptr;
```
Comment (jumaffre) [inline src/node/rpc/handlerregistry.h]: AFAICS, both `consensus` and `history` are only needed by the child class `CommonHandlerRegistry`?

File: src/node/rpc/handlerregistry.h [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+      {
+        return &default_handler.value();
+      }
+
+      return nullptr;
+    }
+
+    virtual void tick(std::chrono::milliseconds elapsed, size_t tx_count) {}
+
+    virtual std::optional<CallerId> valid_caller(
```
Comment (jumaffre) [inline src/node/rpc/handlerregistry.h]: My first feeling is that this should belong to a different class as this feature is not related to the handlers installed? Or go back to `frontend.h`?

**Comment by jumaffre** [comment]
This is awesome! I believe these changes should be reflected in our docs wrt how developers write CCF apps? E.g. https://microsoft.github.io/CCF/developers/logging_cpp.html#rpc-handler

**Comment by cimetrics** [comment]
abstract_context@[3902 aka 20200120.20](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3902&view=results) vs master ewma over 20 builds from [3744](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3744&view=results) to [3900](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3900&view=results)
images

File: src/enclave/rpccontext.h [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+
+    virtual std::vector<uint8_t> serialise_response() const = 0;
+
+    virtual std::vector<uint8_t> result_response(
+      const nlohmann::json& result) const = 0;
+
+    virtual std::vector<uint8_t> error_response(
+      int error, const std::string& msg = "") const = 0;
+
+    virtual void set_response_metafield(
```
Comment (achamayou) [inline src/enclave/rpccontext.h]: header

File: src/enclave/rpccontext.h (hunk 3 above)
Comment (eddyashton) [reply src/enclave/rpccontext.h]: I think we should use `Rpc` to avoid `CCFJSONRPC`. I think I've used `Rpc` in all the new/moved things in this PR, and we can look at a unification pass in future if we don't reach consensus more naturally.

File: src/apps/jsgeneric/jsgeneric.cpp [hunk 7]
```diff
@@ -64,24 +73,45 @@ namespace ccfapp
...
         else
         {
-          return make_pair(false, *err_it);
+          int err_code = jsonrpc::CCFErrorCodes::SCRIPT_ERROR;
+          std::string msg = "";
+
+          if (err_it->is_object())
+          {
+            auto err_code_it = err_it->find("code");
```
[the remark this answers is not in the document]
Comment (eddyashton) [reply src/apps/jsgeneric/jsgeneric.cpp]: I think they'll go away in future, once we have #705. So instead of this `"code"` needing to match the name of a field in a helper in the `__environment` script, as we currently do, we'll just be able to call `ctx.set_error(code, msg)` (or similar) directly from Lua/JS.

File: src/node/rpc/handlerregistry.h (hunk 4 above)
Comment (eddyashton) [reply src/node/rpc/handlerregistry.h]: True, but they're _set_ through this generic `HandlerRegistry` so I think its cleaner to store them here (rather than `virtual set_history(History*) {}`). This is another rough area that I will hope will polish up into something cleaner in future (briefly - only a few of the handlers need these, so it should only be visible to them).

File: src/node/rpc/handlerregistry.h (hunk 5 above)
Comment (eddyashton) [reply src/node/rpc/handlerregistry.h]: I wrote a big blurb but didn't convince myself, so here's a short cop-out answer:

I'd like to pull as much out of `frontend.h` as possible, so that we have greater freedom to implement dispatch, auth, and validation as far down the stack as possible, in potentially handler-specific ways. In practice we don't have a use-case for handler-specific `valid_caller` yet, so this may make more sense back in `frontend.h` in future. But for now I'm happy to pull too many things out rather than too few, in the assumption that these will be tidied in a future pass.

File: src/enclave/enclavetypes.h (hunk 2 above)
Comment (eddyashton) [reply src/enclave/enclavetypes.h]: Done, went with `forwardertypes.h`. There's scope for a big rewrite of the names around here - the dependencies between `enclave/`, `node/` and `node/rpc/` are increasingly unclear.

**Comment by eddyashton** [comment]
This resolves #682.

**Comment by eddyashton** [comment]
> This is awesome! I believe these changes should be reflected in our docs wrt how developers write CCF apps? E.g. https://microsoft.github.io/CCF/developers/logging_cpp.html#rpc-handler

Avoiding documenting right now for what is hopefully an ugly inter-release state, will finalise or write-up before v0.7, ticketed in #706.

**Comment by cimetrics** [comment]
abstract_context@[3904 aka 20200120.21](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3904&view=results) vs master ewma over 20 builds from [3744](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3744&view=results) to [3900](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=3900&view=results)
images
